File: cpp/color_wheel.cpp
#include "color_wheel.hpp"

using namespace WarGrey::STEM;

/*************************************************************************************************/
static const size_t color_count = 36U;
static const float color_radius = 16.0F;
static const float wheel_radius = 256.0F;

static const float color_mixture_radius = 80.0F;

/*************************************************************************************************/
void WarGrey::STEM::ColorWheelWorld::load(float width, float height) {
    float delta_deg = 360.0F / float(color_count);

    this->set_background(0x000000U);

    for (float deg = 0.0F; deg < 360.0F; deg += delta_deg) {
        this->colors.push_back(this->insert(new Circlet(color_radius, deg)));
    }

    this->color_components.push_back(this->insert(new Ellipselet(color_mixture_radius, 0xFF0000U)));
    this->color_components.push_back(this->insert(new Ellipselet(color_mixture_radius, 0x00FF00U)));
    this->color_components.push_back(this->insert(new Ellipselet(color_mixture_radius, 0x0000FFU)));

    for (auto com : this->color_components) {
        com->set_color_mixture(ColorMixture::Add);
    }

    TheBigBang::load(width, height);

    this->tooltip = this->insert(make_label_for_tooltip(GameFont::Tooltip()));
    this->set_tooltip_matter(this->tooltip);
}

void WarGrey::STEM::ColorWheelWorld::reflow(float width, float height) {
    float delta_deg = 360.0F / float(color_count);
    float cx = width * 0.5F;
    float cy = height * 0.5F;
    float x, y;
    int idx = 0;

    TheBigBang::reflow(width, height);
    
    for (float deg = 0.0F; deg < 360.0F; deg += delta_deg) {
        circle_point(wheel_radius, deg - 90.0F, &x, &y, false);
        this->move_to(this->colors[idx++], cx + x, cy + y, MatterAnchor::CC);
    }

    this->move_to(this->color_components[1], 0.0F, height, MatterAnchor::LB);
    this->move_to(this->color_components[2], this->color_components[1], MatterAnchor::CC, MatterAnchor::LC);
    this->move_to(this->color_components[0], this->color_components[1], MatterAnchor::RC, MatterAnchor::CB, -color_mixture_radius * 0.5F);
}

void WarGrey::STEM::ColorWheelWorld::after_select(IMatter* m, bool yes) {
    if (yes) {
        auto com = dynamic_cast<Circlet*>(m);

        if (com != nullptr) {
            this->color_components[this->selection_seq]->set_color(com->get_color());
            this->selection_seq = (this->selection_seq + 1) % this->color_components.size();
        }
    }
}

bool WarGrey::STEM::ColorWheelWorld::update_tooltip(IMatter* m, float x, float y) {
    bool updated = false;
    auto com = dynamic_cast<Circlet*>(m);

    if (com != nullptr) {
        this->tooltip->set_text(" Hue: %.2f ", com->get_body_hsb_hue());
        updated = true;
    }

    return updated;
}

File: python/ColorWheel.py
#!/usr/bin/env python3

from digitama.big_bang.game import *    # 导入游戏模块，内含 Plane 类和常用函数

###############################################################################
# 本游戏中的常量设定
color_count = 36
color_radius = 16.0
wheel_radius = 360.0

mixture_radius: float = 100.0

###############################################################################
class ColorWheelWorld(Plane):
    def __init__(self):
        # 通过父类的构造函数设置窗口标题和帧频
        super(ColorWheelWorld, self).__init__("色相环")

        # 本游戏世界有以下物体
        self.colors = []
        self.color_components = []

    def load(self, Width, Height):
        # 思考：为什么背景一定要是黑色的？
        self.set_background(0x000000)

        self.color_components.append(self.insert(Ellipselet(mixture_radius, mixture_radius, 0xFF0000))) # 红色
        self.color_components.append(self.insert(Ellipselet(mixture_radius, mixture_radius, 0x00FF00))) # 绿色
        self.color_components.append(self.insert(Ellipselet(mixture_radius, mixture_radius, 0x0000FF))) # 蓝色
   
        # 设置混色模式，RGB 属加色模型
        for cc in self.color_components:
            cc.set_color_mixture(ColorMixture.Add)

        self.__load_color_wheel_components()

    def reflow(self, width, height):
        cx = width * 0.5
        cy = height * 0.5
        delta_deg = 360.0 / float(color_count)
        deg = 0.0
        idx = 0

        while deg < 360.0:
            x, y = circle_point(wheel_radius, deg - 90.0, False)
            self.move_to(self.colors[idx], (cx + x, cy + y), MatterAnchor.CC)

            deg += delta_deg
            idx += 1

        self.__reflow_color_components(width * 0.5, height * 0.5)

    def can_select(self, matter):
        return isinstance(matter, Circlet)

    # 实现 ColorMixtureWorld::after_select 方法
    def after_select(self, matter, yes):
        pass

# private
    def __load_color_wheel_components(self):
        delta_deg = 360.0 / float(color_count)
        deg = 0.0

        while deg < 360.0:
            self.colors.append(self.insert(Circlet(color_radius, 0xFF00FF)))
            deg += delta_deg

    def __reflow_color_components(self, x, y):
        cc_off = mixture_radius * 0.5

        self.move_to(self.color_components[0], (x, y), MatterAnchor.CB, 0.0, cc_off)
        self.move_to(self.color_components[1], (self.color_components[0], MatterAnchor.CB), MatterAnchor.RC, cc_off)
        self.move_to(self.color_components[2], (self.color_components[1], MatterAnchor.CC), MatterAnchor.LC)

###############################################################################
launch_universe(ColorWheelWorld, __name__, (1200, 1200))
